#1575 - Fixed creation of vPools concurrently

kvanhijf opened the pull request.

kvanhijf committed:
#1575 - Fixed creation of vPools concurrently

khenderick commented on ovs/extensions/services/systemd.py:
```diff
@@ -208,7 +208,14 @@ def start_service(name, client):
-        return output
+        counter = 0
+        while counter < 10:
+            if Systemd.get_service_status(name=name, client=client)[0] is True:
```
Can you verify why exactly that happened? Did the process log something (e.g. a crash)?

kvanhijf replied:
Process was in `activating` status instead of the expected `active` status

khenderick commented on ovs/lib/mdsservice.py:
```diff
@@ -72,6 +72,9 @@ def prepare_mds_service(storagerouter, vpool, fresh_only):
         from ovs.lib.storagedriver import StorageDriverController
 
+        prepare_mds_lock = volatile_mutex('prepare_mds_{0}'.format(storagerouter.guid))
+        prepare_mds_lock.acquire(30)
```
This must be in a try/finally. Otherwise it might be possible that the lock isn't released.

kvanhijf replied:
Done

khenderick commented on ovs/lib/storagerouter.py:
```diff
@@ -285,6 +285,15 @@ def add_vpool(parameters):
+        # When 2 or more jobs simultaneously run on the same StorageRouter, we need to check and create the StorageDriver partitions in locked context
+        partitions_lock = volatile_mutex('add_vpool_partitions_{0}'.format(storagerouter.guid))
+        try:
+            partitions_lock.acquire(wait=120)
```
Must be in a try/finally or with statement

kvanhijf replied:
Done

khenderick commented on ovs/lib/storagerouter.py:
```diff
@@ -632,7 +587,54 @@ def add_vpool(parameters):
+        except:
+            for sd_partition in storagedriver.partitions:
+                sd_partition.delete()
+            storagedriver.delete()
```
Can you check whether this is the right thing to do? Maybe we can be smarter and cleanup the mess to we have a RUNNING (or none) vPool left.

kvanhijf replied:
Done

kvanhijf committed:
#1575 - Concurrent vPool creation
Changes in ServiceManagement
Start service raises if service is not started properly
Stop service raises if service is not started properly
Get service status only returns the status as returned by systemctl, eg: 'active'

kvanhijf committed:
#1575 - Review feedback
